File: sdks/official/fraiseql-csharp/tests/FraiseQL.Tests/RestAnnotationTests.cs
using FraiseQL.Builders;
using FraiseQL.Registry;
using Xunit;

namespace FraiseQL.Tests;

[Collection(RegistryTestCollection.Name)]
public sealed class RestAnnotationTests : IDisposable
{
    public RestAnnotationTests() => SchemaRegistry.Instance.Clear();
    public void Dispose() => SchemaRegistry.Instance.Clear();

    [Fact]
    public void Query_RestPath_EmitsRestBlock()
    {
        var query = QueryBuilder.Query("users")
            .ReturnType("User")
            .ReturnsList()
            .SqlSource("v_user")
            .RestPath("/api/users")
            .RestMethod("GET")
            .Build();

        Assert.NotNull(query.Rest);
        Assert.Equal("/api/users", query.Rest.Path);
        Assert.Equal("GET", query.Rest.Method);
    }

    [Fact]
    public void Query_RestPath_DefaultsToGet()
    {
        var query = QueryBuilder.Query("users")
            .ReturnType("User")
            .ReturnsList()
            .SqlSource("v_user")
            .RestPath("/api/users")
            .Build();

        Assert.NotNull(query.Rest);
        Assert.Equal("GET", query.Rest.Method);
    }

    [Fact]
    public void Query_NoRestPath_OmitsRestBlock()
    {
        var query = QueryBuilder.Query("users")
            .ReturnType("User")
            .ReturnsList()
            .SqlSource("v_user")
            .Build();

        Assert.Null(query.Rest);
    }

    [Fact]
    public void Mutation_RestPath_EmitsRestBlock()
    {
        var mutation = MutationBuilder.Mutation("createUser")
            .ReturnType("User")
            .SqlSource("fn_create_user")
            .Operation("insert")
            .RestPath("/api/users")
            .RestMethod("POST")
            .Build();

        Assert.NotNull(mutation.Rest);
        Assert.Equal("/api/users", mutation.Rest.Path);
        Assert.Equal("POST", mutation.Rest.Method);
    }

    [Fact]
    public void Mutation_RestPath_DefaultsToPost()
    {
        var mutation = MutationBuilder.Mutation("createUser")
            .ReturnType("User")
            .SqlSource("fn_create_user")
            .Operation("insert")
            .RestPath("/api/users")
            .Build();

        Assert.NotNull(mutation.Rest);
        Assert.Equal("POST", mutation.Rest.Method);
    }

    [Fact]
    public void Mutation_NoRestPath_OmitsRestBlock()
    {
        var mutation = MutationBuilder.Mutation("createUser")
            .ReturnType("User")
            .SqlSource("fn_create_user")
            .Operation("insert")
            .Build();

        Assert.Null(mutation.Rest);
    }

    [Fact]
    public void RestMethod_CaseInsensitive()
    {
        var query = QueryBuilder.Query("users")
            .ReturnType("User")
            .ReturnsList()
            .SqlSource("v_user")
            .RestPath("/api/users")
            .RestMethod("post")
            .Build();

        Assert.NotNull(query.Rest);
        Assert.Equal("POST", query.Rest.Method);
    }
}

File: sdks/official/fraiseql-fsharp/tests/FraiseQL.Tests/RestAnnotationTests.fs
module FraiseQL.Tests.RestAnnotationTests

open Xunit
open FraiseQL

[<Fact>]
let ``Query with restPath emits rest fields in definition`` () =
    let def =
        QueryBuilder.query "users"
        |> QueryBuilder.returnType "User"
        |> QueryBuilder.returnsList true
        |> QueryBuilder.sqlSource "v_user"
        |> QueryBuilder.restPath "/api/users"
        |> QueryBuilder.restMethod "GET"
        |> QueryBuilder.toDefinition

    Assert.Equal(Some "/api/users", def.rest_path)
    Assert.Equal(Some "GET", def.rest_method)

[<Fact>]
let ``Query without restPath has None`` () =
    let def =
        QueryBuilder.query "users"
        |> QueryBuilder.returnType "User"
        |> QueryBuilder.returnsList true
        |> QueryBuilder.sqlSource "v_user"
        |> QueryBuilder.toDefinition

    Assert.Equal(None, def.rest_path)
    Assert.Equal(None, def.rest_method)

[<Fact>]
let ``Mutation with restPath emits rest fields`` () =
    let def =
        MutationBuilder.mutation "createUser"
        |> MutationBuilder.returnType "User"
        |> MutationBuilder.sqlSource "fn_create_user"
        |> MutationBuilder.operation "insert"
        |> MutationBuilder.restPath "/api/users"
        |> MutationBuilder.restMethod "POST"
        |> MutationBuilder.toDefinition

    Assert.Equal(Some "/api/users", def.rest_path)
    Assert.Equal(Some "POST", def.rest_method)

[<Fact>]
let ``Mutation without restPath has None`` () =
    let def =
        MutationBuilder.mutation "createUser"
        |> MutationBuilder.returnType "User"
        |> MutationBuilder.sqlSource "fn_create_user"
        |> MutationBuilder.operation "insert"
        |> MutationBuilder.toDefinition

    Assert.Equal(None, def.rest_path)
    Assert.Equal(None, def.rest_method)
